### ISSUE TITLE
How are the models pretrained ?

### ISSUE THREAD

**Issue by lyxok1** [LoyoYang/DeCoTa#3]
Thanks for sharing the code of this work. Recently I am trying to run the code with default setting, and find before starting, the code will load the pretrained models under `pretrained_models` folder in default by this [code](https://github.com/LoyoYang/DeCoTa/blob/51167d6cffd2bbc4a172f5fbeb7926b521638f90/main.py#L132). When I try to take the original ImageNet pretraining weight of torch, I found the results on Real-Clipart 3 shot varies obviously.

When training with provided pretrained model, the results ended up with test acc `77.4` after 20000 steps of training, while when training from an ImageNet pretraining model, the results can only achieve `70.1`.

I noticed that in the implementation detail of the paper, the two networks are first pretrained with labeled data, but did not find detailed information about the pretraining process, can you provide any suggustion on the setting of pretraining (e.g. starting model, training steps, learning rate), thanks~

**Comment by LoyoYang** [comment]
Thanks for the question. @lyxok1 

As put in Section 4: Implementation Details, "We initialize w_f with a model first finetuned on DS, and initialize w_g with a model first fine-tuned on DS and then fine-tuned on DT. We do so to encourage the two models to be different at the beginning."

You can keep the same learning rate and optimizer, 10,000 steps should be enough for DS and DT.
Let me know if there is still an issue.

**Comment by davidshumway** [comment]
This is my interpretation which includes some tutorial code taken from here: (https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html).

Helper functions:
- train_model()
- setup_data()
```python
import copy, time
def train_model(model, dataloaders, optimizer_f, optimizer_g, num_epochs=25, init_lr=0.001):
  since = time.time()
  val_acc_history = []
  best_model_wts = copy.deepcopy(model.state_dict())
  best_acc = 0.0
  criterion = nn.CrossEntropyLoss()

  for epoch in range(num_epochs):
    print('Epoch {}/{}'.format(epoch, num_epochs - 1))

    # Each epoch has a training and validation phase
    for phase in ['train', 'val21']:
      if phase == 'train':
        model.train()  # Set model to training mode
      else:
        model.eval()   # Set model to evaluate mode

      running_loss = 0.0
      running_corrects = 0

      # Iterate over data.
      print(f'Iterate over data {phase}, batch:', end=' ')
      param_lr_f = []
      param_lr_g = []
      for param_group in optimizer_f.param_groups:
          param_lr_f.append(param_group["lr"])
      for param_group in optimizer_g.param_groups:
          param_lr_g.append(param_group["lr"])
      step = 0

      for inputs, labels in dataloaders[phase]:
        inputs = inputs.to(device)
        labels = labels.to(device)
        optimizer_f.zero_grad()
        optimizer_g.zero_grad()
        if step % 20 == 0:
          print(step, sep='', end=' ')
        step += 1
        # forward
        # track history if only in train
        with torch.set_grad_enabled(phase == 'train'):
          outputs = model(inputs)
          loss = criterion(outputs, labels)
          _, preds = torch.max(outputs, 1)
          # backward + optimize only if in training phase
          if phase == 'train':
            loss.backward()
            optimizer_f.step()
            optimizer_g.step()
            optimizer_f = inv_lr_scheduler(param_lr_f, optimizer_f, step, init_lr=init_lr)
            optimizer_g = inv_lr_scheduler(param_lr_g, optimizer_g, step, init_lr=init_lr)
            

        # statistics
        running_loss += loss.item() * inputs.size(0)
        running_corrects += torch.sum(preds == labels.data)

      epoch_loss = running_loss / len(dataloaders[phase].dataset)
      epoch_acc = running_corrects.double() / len(dataloaders[phase].dataset)
      print('{} Loss: {:.4f} Acc: {:.4f}'.format(phase, epoch_loss, epoch_acc))

      # deep copy the model
      if phase == 'val21' and epoch_acc > best_acc:
        best_acc = epoch_acc
        best_model_wts = copy.deepcopy(model.state_dict())
      if phase == 'val21':
        val_acc_history.append(epoch_acc)

  time_elapsed = time.time() - since
  print('Training complete in {:.0f}m {:.0f}s'.format(time_elapsed // 60, time_elapsed % 60))
  print('Best val Acc: {:4f}'.format(best_acc))

  # load best model weights
  model.load_state_dict(best_model_wts)
  return model, val_acc_history

def setup_data(dataset=None, is_src=True):
  input_size = 224

  data_transforms = {
    'train': transforms.Compose([
      transforms.RandomResizedCrop(input_size),
      transforms.RandomHorizontalFlip(),
      transforms.ToTensor(),
      transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ]),
    'val21': transforms.Compose([
      transforms.Resize(input_size),
      transforms.CenterCrop(input_size),
      transforms.ToTensor(),
      transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ]),
  }

  data_dir = f'./data/multi/{dataset}/'
  d = ['train', 'val21']

  if is_src is True:
    # src train labeled
    image_datasets = {x:
      datasets.ImageFolder(os.path.join(data_dir, x), data_transforms[x])
      for x in d
    }
  else:
    # tgt train labeled (6 samples)
    image_datasets = {
      'train': datasets.ImageFolder(f'./data/multi/{dataset}/train_tgt/', data_transforms['train']),
      'val21': datasets.ImageFolder(f'./data/multi/{dataset}/val21', data_transforms['val21'])
    }

  # Dataloaders
  # bs=24 per https://github.com/LoyoYang/DeCoTa/blob/
  #   d75146f1d8b5e12718edab8ff4c1eb442506a14b/utils/return_dataset.py#L89
  dataloaders_dict = {x:
      torch.utils.data.DataLoader(image_datasets[x], batch_size=24, shuffle=True, num_workers=2)
    for x in d
  }

  return dataloaders_dict 
```

For three domains, create and save pre-trained models for each domain. For example, for a single Src/Tgt combination, there are two generated models:
* `pretrained_src_{src}_to_{tgt}.pth.tar`
* `pretrained_tgt_{src}_to_{tgt}.pth.tar`

For each Src, there are thus 4 models (src1 src1_to_tgt1, src1 src1_to_tgt2).

In total, there are 4x3=12 models:
```python
from torchvision import datasets, models, transforms
import torch
import os
import torch.utils.model_zoo as model_zoo
import torch.nn as nn
from model.resnet_model import ResNet, BasicBlock, model_urls, resnet34, MetaResnet34
import torch.optim as optim
from utils.lr_schedule import inv_lr_scheduler

domains = ['SB', 'NB', 'CB']
init_lr = 0.001

for src in domains:
  net = MetaResnet34(num_class=2)
  params_1 = []
  for value in net.G.params():
    if value.requires_grad:
      params_1 += [{'params': [value], 'lr': init_lr,
        'weight_decay': 0.0005}]
  params_F1 = []
  for value in net.F1.params():
    if value.requires_grad:
      params_F1 += [{'params': [value], 'lr': init_lr,
        'weight_decay': 0.0005}]

  device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
  net = net.to(device)
  
  optimizer_f = optim.SGD(params_1, lr=init_lr, momentum=0.9)
  optimizer_g = optim.SGD(params_F1, lr=init_lr, momentum=0.9)

  # First model fine-tuned on labeled src
  print('training src model')
  model, hist = train_model(net, setup_data(dataset=src, is_src=True), optimizer_f,
    optimizer_g, num_epochs=1, init_lr=init_lr)
    
  for tgt in domains:
    if src == tgt: continue
    print(f'src{src} tgt{tgt}')
    
    # Save the first model
    torch.save(model.state_dict(), f'./pretrained_models/pretrained_src_{src}_to_{tgt}.pth.tar')

    # Second model further fine-tuned on 6 labeled tgt samples
    print('training src+tgt model')
    model2, hist = train_model(copy.copy(model), setup_data(dataset=tgt, is_src=False),
      optimizer_f, optimizer_g, num_epochs=1, init_lr=init_lr)
    torch.save(model2.state_dict(), f'./pretrained_models/pretrained_tgt_{src}_to_{tgt}.pth.tar')

```

I used this for a custom dataset with domins named "SB", "NB", and "CB."

The number of files in each domain are as follows:

```
labeled_source_images_SB 2444
labeled_target_images_SB_3 6
unlabeled_target_images_SB_3 2438
validation_target_images_SB_3 703
labeled_source_images_NB 2464
labeled_target_images_NB_3 6
unlabeled_target_images_NB_3 2458
validation_target_images_NB_3 601
labeled_source_images_CB 1396
labeled_target_images_CB_3 6
unlabeled_target_images_CB_3 1390
validation_target_images_CB_3 394
```

The training time for 10 epochs is around 20 minutes using a single GPU. The problem is a binary two-class classification problem.

**Comment by davidshumway** [comment]
Note the above is for binary classification so `criterion` could be set to `nn.BCELoss()`. Then 

```
criterion = nn.BCELoss()
loss = criterion(outputs, labels.unsqueeze(1))
preds = torch.round(outputs).to(int).flatten()
```
